3p/ampcontext.js

```diff
@@ -0,0 +1,203 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import './polyfills';
+import {
+  dev,
+  user,
+} from '../src/log';
+import {IframeMessagingClient} from './iframe-messaging-client';
+
+/**
+   Enum for the different postmessage types for the window.context
+   postmess api.
+*/
+export const MessageType_ = {
+  SEND_EMBED_STATE: 'send-embed-state',
+  EMBED_STATE: 'embed-state',
+  SEND_EMBED_CONTEXT: 'send-embed-context',
+  EMBED_CONTEXT: 'embed-context',
+  SEND_INTERSECTIONS: 'send-intersections',
+  INTERSECTION: 'intersection',
+  EMBED_SIZE: 'embed-size',
+  EMBED_SIZE_CHANGED: 'embed-size-changed',
+  EMBED_SIZE_DENIED: 'embed-size-denied',
+};
+
+export class AmpContext extends IframeMessagingClient {
+
+  /**
+   *  @param {Window} win The window that the instance is built inside.
+   */
+  constructor(win) {
+    super(win);
+    this.setupMetadata_();
+    this.ampWindow = this.getHostWindow();
+  }
+
+  /** @override */
+  getHostWindow() {
+    const sentinelMatch = this.sentinel.match(/((\d+)-\d+)/);
+    if (sentinelMatch) {
+      this.depth = Number(sentinelMatch[2]);
+      this.ancestors = [];
+      for (let win = this.win_; win && win != win.parent; win = win.parent) {
+        // Add window keeping the top-most one at the front.
+        this.ancestors.unshift(win.parent);
+      }
+      return this.ancestors[this.depth];
+    } else {
+      dev().error('AMPCONTEXT', '- Incorrect sentinel format.');
+      throw new Error('Incorrect sentinel format.');
+    }
+  }
+
+  /** @override */
+  getSentinel() {
+    return this.sentinel;
+  }
+
+  /** @override */
+  registerCallback_(messageType, callback) {
+    if (!this.isValidMessageType(messageType)) {
+      user().error('AMPCONTEXT', '- Invalid message type.');
+      return;
+    }
+    this.callbackFor_[messageType] = callback;
+    return () => { delete this.callbackFor_[messageType]; };
+  }
+
+
+  /**
+   *  Return true if messagetype is approved
+   */
+  isValidMessageType(messageType) {
+    const validTypes = [
+      'embed-state',
+      'embed-context',
+      'intersection',
+      'embed-size',
+      'embed-size-changed',
+      'embed-size-denied',
+    ];
+    return !!(validTypes.includes(messageType));
+  }
+
+  /**
+   *  Send message to runtime to start sending page visibility messages.
+   *  @param {function(Object)} callback Function to call every time we receive a
+   *    page visibility message.
+   *  @returns {function()} that when called stops triggering the callback
+   *    every time we receive a page visibility message.
+   */
+  observePageVisibility(callback) {
+    const stopObserveFunc = this.registerCallback_(MessageType_.EMBED_STATE,
+                                                   callback);
```

Comment:
we use 4 space indentation for line breaks.

in this case, you can break after '('